presto-main/src/main/java/com/facebook/presto/sql/planner/CanonicalPlanHashes.java

```diff
@@ -0,0 +1,69 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.planner;
+
+import com.facebook.presto.common.plan.PlanCanonicalizationStrategy;
+import com.facebook.presto.spi.PrestoException;
+import com.facebook.presto.spi.plan.PlanNode;
+import com.facebook.presto.spi.plan.PlanNodeId;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableMap;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import static com.facebook.presto.spi.StandardErrorCode.PLAN_SERIALIZATION_ERROR;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.hash.Hashing.sha256;
+import static java.nio.charset.StandardCharsets.UTF_8;
+
+public class CanonicalPlanHashes
+{
+    private final Map<PlanNodeId, Map<PlanCanonicalizationStrategy, String>> hashes = new HashMap<>();
```

Comment:
I don't love this current structure (have a place holder with this mutable list in the session because it's a convenient spot, and then build it later).  And it's mutable without any protections. It would be much better to create this object only when we need it, and then have the map be immutable.

Another option might be to add the field directly to the plan Node instead of the canonicalId.  We'd want to null it out before sending to the workers though. 

another option (which needs a bit more thought to ensure it doesn't have negative GC implications), would be instead of having a canonicalId and then these canonicalHashes that get passed around, instead add a self reference in the planNode as a field "statsEquivalentPlanNode" or something.  When you change the plan during optimization, keep passing the same statsEquivalentPlanNode (similar to how canonicalId is used now) And then don't to any hashing here at all.  Then later if we need stats during optimization, then we hash the "statsEquivalentPlanNode".  Prevents us from doing unnecessary work since we only hash as needed, doesn't grow exponentially with the size of the plan tree, etc.  We'd ideally still null it out when passing the plan to the workers.

Reply:
how big can this get, given that it's got the plan tree repeated a bunch of times at different levels?  Should we be worried about memory usage?  Also, the more strategies we add the larger this will grow.

Reply:
Going for Storing plan node id for now, as it reveals less details in the spi interface.

The analysis of 2 options:

1) Store stats equivalent plan node id. This will lead to twice memory used by plan nodes. This memory will last for lifetime of query rather than lifetime of planning, unless we remove it again after planning is done. This will also try to canonicalize plan nodes multiple times, and we may need some kind of caching/precomputing mechanism for that.

2) Store stats equivalent plan node hashes for all plan nodes. We can do the precomputation in a single step. As we store sha256 hashes, memory used in minimal. Storing hashes for all plan nodes may seem unnecessary, but a lot of it is going to be used anyway. I still think overhead may be better than the overhead above.

Let me know what you think!

Reply:
I like this approach.  I agree we probably will want some kind of caching still for the plan nodes, but this at least makes the interface really nice.  When testing we can look out for issues with the plan being too big, and whether we should null this out for the plan sent to workers.